gradio/route_utils.py

```diff
@@ -0,0 +1,248 @@
+                return True
+        return False
+
+    def keys(self):
+        return self.__dict__.keys()
+
+    def values(self):
+        return self.__dict__.values()
+
+    def items(self):
+        return self.__dict__.items()
+
+    def __str__(self) -> str:
+        return str(self.__dict__)
+
+    def __repr__(self) -> str:
+        return str(self.__dict__)
+
+
+@document()
+class Request:
+    """
+    A Gradio request object that can be used to access the request headers, cookies,
+    query parameters and other information about the request from within the prediction
+    function. The class is a thin wrapper around the fastapi.Request class. Attributes
+    of this class include: `headers`, `client`, `query_params`, and `path_params`. If
+    auth is enabled, the `username` attribute can be used to get the logged in user.
+    Example:
+        import gradio as gr
+        def echo(name, request: gr.Request):
+            print("Request headers dictionary:", request.headers)
+            print("IP address:", request.client.host)
+            return name
+        io = gr.Interface(echo, "textbox", "textbox").launch()
+    """
+
+    def __init__(
+        self,
+        request: fastapi.Request | None = None,
+        username: str | None = None,
+        **kwargs,
+    ):
+        """
+        Can be instantiated with either a fastapi.Request or by manually passing in
+        attributes (needed for websocket-based queueing).
+        Parameters:
+            request: A fastapi.Request
+        """
+        self.request = request
+        self.username = username
+        self.kwargs: dict = kwargs
+
+    def dict_to_obj(self, d):
+        if isinstance(d, dict):
+            return json.loads(json.dumps(d), object_hook=Obj)
+        else:
+            return d
+
+    def __getattr__(self, name):
+        if self.request:
+            return self.dict_to_obj(getattr(self.request, name))
+        else:
+            try:
+                obj = self.kwargs[name]
+            except KeyError as ke:
+                raise AttributeError(
+                    f"'Request' object has no attribute '{name}'"
+                ) from ke
+            return self.dict_to_obj(obj)
+
+
+class FnIndexInferError(Exception):
+    pass
+
+
+def infer_fn_index(app: App, api_name: str, body: PredictBody) -> int:
+    if body.fn_index is None:
+        for i, fn in enumerate(app.get_blocks().dependencies):
+            if fn["api_name"] == api_name:
+                return i
+
+        raise FnIndexInferError(f"Could not infer fn_index for api_name {api_name}.")
+    else:
+        return body.fn_index
+
+
+def compile_gr_request(
+    app: App,
+    body: PredictBody,
+    fn_index_inferred: int,
+    username: Optional[str],
+    request: Optional[fastapi.Request],
+):
+    # If this fn_index cancels jobs, then the only input we need is the
+    # current session hash
+    if app.get_blocks().dependencies[fn_index_inferred]["cancels"]:
```

Comment:
Thank you! This is a critical point. I will take a look into it...

Reply:
### Reason
This line is what cancels the task and the cause of that `CancelledError`:
https://github.com/gradio-app/gradio/blob/3d66e61d641da8ca2a7d10c545c7dc0139697f00/gradio/routes.py#L655

The cancelled task here is:
1. At first, set at `Queue.start_processing()`.
   https://github.com/gradio-app/gradio/blob/3d66e61d641da8ca2a7d10c545c7dc0139697f00/gradio/queueing.py#L146
2. Then repeatedly renewed just after cancelling the existing one with the same task name:
   https://github.com/gradio-app/gradio/blob/3d66e61d641da8ca2a7d10c545c7dc0139697f00/gradio/routes.py#L658-L660

and even in the main branch code, the `CancelledError` has been raised when the task is cancelled, but there was no `try-except` block to catch it.
(For example, you can catch the error by adding `except asyncio.CancelledError` to the following try-block:
https://github.com/gradio-app/gradio/blob/3d66e61d641da8ca2a7d10c545c7dc0139697f00/gradio/queueing.py#L391
)

What changed is this `except BaseException` handler:
https://github.com/gradio-app/gradio/pull/5165/files#diff-cab37b7b6ce45220156832fd4ea1597bedcb26f5cfc43ab5c678fc8a586a0232R398
I just copied it [from `routes.py`](https://github.com/gradio-app/gradio/blob/3d66e61d641da8ca2a7d10c545c7dc0139697f00/gradio/routes.py#L538) to do the same thing in `queueing.py`, but it accidentally worked as a handler for the `CancelledError` and prints the stack trace with `traceback.print_exc()`.

### Solution
To ignore the `CancelledError` keeping the same behavior as before for other exceptions, we can just add `except asyncio.CancelledError` to the `try-except` block:
```diff
                 iterators=iterators,
                 fn_index_inferred=fn_index_inferred,
             )
+        except asyncio.CancelledError:
+            raise
         except BaseException as error:
             show_error = app.get_blocks().show_error or isinstance(error, Error)
             traceback.print_exc()
```

However, as a general practice, `except Exception` is preferred to `except BaseException`:
> Exceptions which are not subclasses of Exception are not typically handled, because they are used to indicate that the program should terminate.
> https://docs.python.org/3/tutorial/errors.html#handling-exceptions

So replacing `except BaseException` with `except Exception` looks better if we recall the reason why `BaseException` is used here and can confirm that it's ok to replace it.

Reply:
@freddyaboulton @abidlabs I pushed the fix: 0b2c42c.
However, as commented above, replacing `except BaseException` with `except Exception` might be better instead of this commit, if we can judge it's ok knowing the reason of having been using `BaseException`.
Do you have any idea?

Reply:
Should be fine to replace with `except Exception` -- any objections @freddyaboulton?

Reply:
Thanks @abidlabs and @whitphx !